### PR TITLE
feat(profile): add profile command to surface CPU profiling data

### LINKED ISSUE #56: Add profile command to surface profiling data
## Summary

Add a `profile` command that fetches and displays profiling data from Sentry in a format useful for analysis (both human and LLM consumption).

## Background

From internal discussion: Adding profile support to the CLI would allow LLMs to grep/analyze profiling data.

## Proposed Approach

1. **Fetch profile data** from Sentry API
2. **Convert to a graph structure** and surface the leaf nodes with the most samples
3. **Ignore system frames** to produce cleaner results
4. **Output in a consumable format** (e.g., markdown table, JSON) that works well for both humans and LLM analysis

## Potential Output

- Top functions by sample count (excluding system frames)
- Call hierarchy / flamegraph as markdown
- Hotspot identification

### PR DESCRIPTION
**Pull request by BYK**
## Summary

Adds a new `sentry profile` command with two subcommands to surface CPU profiling data from Sentry:

- **`profile list`** - Lists transactions that have profiling data, with sample counts, p75/p95 durations, and short aliases
- **`profile view <transaction>`** - Analyzes CPU profile for a transaction, showing hot paths and performance metrics

## Features

- Hot path detection with CPU time percentages
- P75/P95/P99 percentile statistics per function
- User code vs library code filtering (`--allFrames` to include all)
- JSON output for CI/automation
- `--web` flag to open profiles in Sentry UI
- Configurable time periods (1h/24h/7d/14d/30d)
- **Transaction aliases** for quick access via numeric index or short alias
- **Smart formatting**: common prefix stripping, middle-truncation for long names
- **SAMPLES column** showing real profile sample counts via `count_unique(timestamp)`

### List Transactions with Profiles

```bash
$ sentry profile list my-org/backend --period 7d

Transactions with Profiles in my-org/backend (last 7d):

  #   ALIAS   TRANSACTION                                         SAMPLES         p75         p95
─────────────────────────────────────────────────────────────────────────────────────────────────────
    1   u       projects/{project_id}/users/                            42        3.8s        5.0s
    2   a       webhooks/provision/account/                             18        2.7s        2.7s
    3   c       organizations/{org_id}/code-mappings/                    6        2.1s        2.1s
    4   e       projects/{project_id}/events/                          291        1.5s        8.6s
    5   i       organizations/{org_id}/issues/                         541        1.5s        2.8s

Common prefix stripped: /api/0/
Tip: Use 'sentry profile view 1' or 'sentry profile view <alias>' to analyze.
```

### View Profile Analysis

Access by **index**, **alias**, or **full transaction name**:

```bash
# By numeric index (from the list)
$ sentry profile view my-org/backend 1

# By short alias
$ sentry profile view my-org/backend e

# By full transaction name
$ sentry profile view my-org/backend "/api/0/projects/{project_id}/events/"
```

Output:
```
/api/0/projects/{project_id}/events/: CPU Profile Analysis (last 7d)
════════════════════════════════════════════════════════════════════════════════

Performance Percentiles
  p75: 1.7s    p95: 12.1s    p99: 12.1s

Hot Paths (Top 10 by CPU time, user code only)
────────────────────────────────────────────────────────────
    #   Function                                  Location                        % Time
    1   EnvMiddleware.<locals>.EnvMiddleware_impl  middleware/env.py:14              7.7%
    2   access_log_middlewa…<locals>.middleware    middlew…ess_log.py:171            7.7%
    3   SubdomainMiddleware.__call__               middlew…ubdomain.py:53            7.7%
    4   AIAgentMiddleware.__call__                 middlew…ai_agent.py:97            7.6%
    5   IntegrationControlMiddleware.__call__      middlew…_control.py:60            7.6%
```

### Stale Alias Detection

If you change the period or project, the CLI warns you:

```bash
$ sentry profile view i --period 24h

Error: Transaction alias 'i' is from a different time period (cached: 7d, requested: 24h).

Suggestion: Run 'sentry profile list my-org/backend --period 24h' to refresh aliases.
```

### Other Options

```bash
# Include library/system frames
sentry profile view my-org/backend 1 --allFrames

# JSON output for automation
sentry profile view my-org/backend 1 --json

# Open in Sentry web UI
sentry profile view my-org/backend 1 --web
```

## How Aliases Work

- Aliases are generated from the **last meaningful URL segment** (e.g., `/api/.../issues/` → `i`)
- Stored in SQLite with fingerprint `org:project:period` for cache validation
- Uses **shortest unique prefix** algorithm (same as project aliases for issues)

## Formatting Improvements

- **SAMPLES column**: Uses `count_unique(timestamp)` which returns real sample counts (not `count()` which returned 0)
- **Common prefix stripping**: Strips shared path prefix (e.g., `/api/0/`) across all transactions for readability
- **Middle-truncation**: Long names are truncated with `…` in the middle, preserving both start and end
- **Location column**: Renamed from "File:Line", widened to 30 chars with middle-truncation
- **No-data error**: Suggests running `profile list` instead of vague "check the transaction name"

Closes #56

### CODE REVIEW

File: src/lib/transaction-alias.ts [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    if (count === 0) {
+      result.push(segment);
+    } else {
+      // Append numeric suffix for duplicates (issues2, issues3, etc.)
+      result.push(`${segment}${count + 1}`);
+    }
+  }
+
+  return result;
+}
```
[the remark this answers is not in the document]
Comment (BYK) [reply src/lib/transaction-alias.ts:125]: Fixed. `disambiguateSegments` now tracks all result values in a `resultSet` and increments the numeric suffix until a unique name is found, preventing collisions with raw segments like `"issues2"`. Added property-based tests in `test/lib/transaction-alias.property.test.ts` to verify the uniqueness invariant holds for all inputs.

File: src/commands/profile/view.ts [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+      `Project "${projectSlug}" exists in multiple organizations`,
+      `sentry profile view <org>/${projectSlug} ${transactionRef}`,
+      alternatives
+    );
+  }
+  const foundProject = found[0];
+  if (!foundProject) {
+    throw new ContextError(`Project "${projectSlug}" not found`, USAGE_HINT);
+  }
+  const orgSlug = foundProject.organization?.slug;
```
[the remark this answers is not in the document]
Comment (BYK) [reply src/commands/profile/view.ts]: Fixed. The entire `resolveFromProjectSearch` function has been replaced by a shared `resolveProjectBySlug()` in `src/lib/resolve-target.ts`, which correctly uses `foundProject.orgSlug` (the reliable property from `ProjectWithOrg`). This also addresses Bug 3 (deduplication).

File: src/commands/profile/view.ts [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+      USAGE_HINT
+    );
+  }
+  return {
+    org: orgSlug,
+    project: foundProject.slug,
+    orgDisplay: orgSlug,
+    projectDisplay: foundProject.slug,
+  };
+}
```
[the remark this answers is not in the document]
Comment (BYK) [reply src/commands/profile/view.ts]: Fixed. Extracted a shared `resolveProjectBySlug()` into `src/lib/resolve-target.ts` and updated all three command files (`event/view.ts`, `log/view.ts`, `profile/view.ts`) to use it. This eliminates the duplication and the bug it caused (using `organization?.slug` instead of `orgSlug`).

File: src/lib/sentry-urls.ts [hunk 4]
```diff
@@ -105,6 +105,39 @@ export function buildBillingUrl(orgSlug: string, product?: string): string {
...
+ * @param transactionName - Transaction name to view profiles for
+ * @returns Full URL to the profiling flamegraph view
+ */
+export function buildProfileUrl(
+  orgSlug: string,
+  projectSlug: string,
+  transactionName: string
+): string {
+  const encodedTransaction = encodeURIComponent(transactionName);
+  return `${getSentryBaseUrl()}/organizations/${orgSlug}/profiling/profile/${projectSlug}/flamegraph/?query=transaction%3A${encodedTransaction}`;
```
[the remark this answers is not in the document]
Comment (BYK) [reply src/lib/sentry-urls.ts]: Fixed. `buildProfileUrl` now wraps the transaction name in encoded quotes: `transaction%3A%22${encodedTransaction}%22`. Updated the property test in `test/lib/sentry-urls.property.test.ts` to verify this.

File: src/commands/profile/list.ts [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+ * Parse and validate the stats period.
+ */
+function parsePeriod(value: string): string {
+  if (!VALID_PERIODS.includes(value)) {
+    throw new Error(
+      `Invalid period. Must be one of: ${VALID_PERIODS.join(", ")}`
+    );
+  }
+  return value;
+}
```
[the remark this answers is not in the document]
Comment (BYK) [reply src/commands/profile/list.ts:102]: Fixed. Extracted `VALID_PERIODS` and `parsePeriod` into `src/commands/profile/shared.ts`. Both `list.ts` and `view.ts` now import from the shared module.

File: src/commands/profile/list.ts [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+      projectSlug = parsed.projectSlug;
+    }
+
+    // Resolve org and project
+    const resolvedTarget = await resolveOrgAndProject({
+      org: parsed.type === "explicit" ? parsed.org : undefined,
+      project: projectSlug,
+      cwd,
+      usageHint: USAGE_HINT,
+    });
```
[the remark this answers is not in the document]
Comment (BYK) [reply src/commands/profile/list.ts]: Fixed. Refactored `profile list` to use a new `resolveListTarget()` function with a proper `project-search` case that calls `findProjectsBySlug` directly (matching the pattern in `log/list.ts` and `profile/view.ts`). Added tests covering single match, multiple matches, and no matches scenarios.

File: src/commands/profile/list.ts [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+        );
+        throw new ContextError(
+          `Project "${parsed.projectSlug}" exists in multiple organizations`,
+          `sentry profile list <org>/${parsed.projectSlug}`,
+          alternatives
+        );
+      }
+      const match = matches[0] as (typeof matches)[0];
+      return { org: match.orgSlug, project: match.slug };
+    }
```
[the remark this answers is not in the document]
Comment (BYK) [reply src/commands/profile/list.ts]: Fixed — `resolveListTarget` now delegates to `resolveProjectBySlug()` for the `project-search` case, eliminating the duplicated `findProjectsBySlug` + error handling logic. Updated the test to expect `ValidationError` (which `resolveProjectBySlug` throws for multi-match) instead of `ContextError`.

File: src/lib/profile/analyzer.ts [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+  if (ms >= 1) {
+    return `${ms.toFixed(2)}ms`;
+  }
+  // Sub-millisecond
+  const us = ms * 1000;
+  if (us >= 1) {
+    return `${us.toFixed(0)}µs`;
+  }
+  return `${(us * 1000).toFixed(0)}ns`;
+}
```
[the remark this answers is not in the document]
Comment (BYK) [reply src/lib/profile/analyzer.ts:72]: Fixed — renamed `formatDuration` → `formatDurationMs` in `analyzer.ts` to disambiguate from the `formatDuration` in `formatters/human.ts` (which takes seconds). Updated all call sites in `formatters/profile.ts` and corresponding test file.

File: src/commands/profile/list.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (BYK) [reply src/commands/profile/list.ts]: Fixed — `resolveListTarget` now delegates to `resolveProjectBySlug()` for the `project-search` case, eliminating the duplicated `findProjectsBySlug` + error handling logic. Updated the test to expect `ValidationError` (which `resolveProjectBySlug` throws for multi-match) instead of `ContextError`.

File: src/lib/profile/analyzer.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (BYK) [reply src/lib/profile/analyzer.ts:72]: Fixed — renamed `formatDuration` → `formatDurationMs` in `analyzer.ts` to disambiguate from the `formatDuration` in `formatters/human.ts` (which takes seconds). Updated all call sites in `formatters/profile.ts` and the test file.

[the remark this answers is not in the document]
Comment (BYK) [reply src/lib/profile/analyzer.ts:72]: Fixed — renamed `formatDuration` → `formatDurationMs` in `analyzer.ts` to disambiguate from the `formatDuration` in `formatters/human.ts` (which takes seconds). Updated all call sites in `formatters/profile.ts` and the test file.

[the remark this answers is not in the document]
Comment (BYK) [reply src/lib/profile/analyzer.ts:72]: Fixed: renamed formatDuration to formatDurationMs in analyzer.ts to disambiguate from the formatDuration in formatters/human.ts (which takes seconds). Updated all call sites in formatters/profile.ts and the test file.

File: src/lib/resolve-transaction.ts [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+/** Pattern to detect numeric-only input */
+const NUMERIC_PATTERN = /^\d+$/;
+
+/**
+ * Check if input is a full transaction name (contains / or .).
+ * Full names are passed through without alias lookup.
+ */
+function isFullTransactionName(input: string): boolean {
+  return input.includes("/") || input.includes(".");
+}
```
[the remark this answers is not in the document]
Comment (BYK) [reply src/lib/resolve-transaction.ts:73]: Fixed — replaced `isFullTransactionName` (which checked for `/` or `.`) with an inverted `isAliasLike` check. Now only short, single-case letter strings (with optional numeric suffix) matching the pattern `^(?:[a-z]+|[A-Z]+)\d*$` are routed to alias lookup. Everything else — including bare names like `process_request`, `handle-webhook`, `ProcessEvent`, `GET /users` — passes through as a full transaction name.

This matches how `buildTransactionAliases` actually generates aliases: lowercased letter-only prefixes with optional numeric disambiguators. Added 6 new tests covering underscored, hyphenated, uppercase, mixed-case, spaced, and colon-containing transaction names.

File: src/lib/profile/analyzer.ts [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
```
[the remark this answers is not in the document]
Comment (BYK) [reply src/lib/profile/analyzer.ts:72]: Fixed — `formatDurationMs` now re-checks after rounding to ensure values that cross a tier boundary are promoted to the correct format:

- `999.5ms`: `Math.round(999.5) = 1000` → now outputs `"1.0s"` instead of `"1000ms"`
- `99.95ms`: `(99.95).toFixed(1) = "100.0"` → now outputs `"100ms"` instead of `"100.0ms"`

Added explicit boundary test cases for both scenarios.

File: src/lib/db/transaction-aliases.ts [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+ * Returns the stale fingerprint if found, null otherwise.
+ */
+export function getStaleIndexFingerprint(idx: number): string | null {
+  const db = getDatabase();
+
+  const row = db
+    .query("SELECT fingerprint FROM transaction_aliases WHERE idx = ? LIMIT 1")
+    .get(idx) as { fingerprint: string } | undefined;
+
+  return row?.fingerprint ?? null;
```
[the remark this answers is not in the document]
Comment (BYK) [reply src/lib/db/transaction-aliases.ts]: Fixed — both `getStaleFingerprint` and `getStaleIndexFingerprint` now accept a `currentFingerprint` parameter and exclude it from the query using `AND fingerprint != ?`. This ensures they only return fingerprints from different contexts, preventing misleading stale-alias error messages.

Added dedicated tests verifying that querying with the same fingerprint returns null.

File: src/lib/transaction-alias.ts [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+    // Normalize: remove hyphens/underscores, lowercase
+    return segment.replace(/[-_]/g, "").toLowerCase();
+  }
+
+  // Fallback: use first non-empty, non-numeric segment if no meaningful one found
+  const firstSegment = segments.find(
+    (s) =>
+      s.length > 0 && !NUMERIC_PATTERN.test(s) && !PLACEHOLDER_PATTERN.test(s)
+  );
+  return firstSegment?.replace(/[-_]/g, "").toLowerCase() ?? "txn";
```
[the remark this answers is not in the document]
Comment (BYK) [reply src/lib/transaction-alias.ts]: Fixed. `extractTransactionSegment` now skips segments that normalize to empty strings (e.g. `"-"`, `"_"`, `"---___"`). The loop continues to the next segment, and the final fallback also guards against empty strings, returning `"txn"` as a safe default.

Added 4 edge-case tests:
- hyphen-only segment → `"txn"`
- underscore-only segment → `"txn"`
- mixed hyphens/underscores → `"txn"`
- skips hyphen-only to find next meaningful segment

File: src/lib/transaction-alias.ts [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+      resultSet.add(candidate);
+    } else {
+      // Raw segment name is available
+      result.push(segment);
+      resultSet.add(segment);
+    }
+  }
+
+  return result;
+}
```
[the remark this answers is not in the document]
Comment (BYK) [reply src/lib/transaction-alias.ts:125]: Fixed. The root cause was that `disambiguateSegments` appended numeric suffixes (`"issues2"`), creating a prefix relationship with the original (`"issues"`). This caused `findShortestUniquePrefixes` to degrade both to full-length strings since no prefix of `"issues"` is unique when `"issues2"` also starts with it.

Changed disambiguation to **prepend `"x"` characters** instead:
- `["issues", "events", "issues"]` → `["issues", "events", "xissues"]`
- Collision handling: `["issues", "xissues", "issues"]` → `["issues", "xissues", "xxissues"]`

This breaks the prefix relationship, so `findShortestUniquePrefixes` now gives `"i"` and `"x"` instead of `"issues"` and `"issues2"`. Also fixed the incorrect docstring example.

Added a dedicated test verifying duplicate segments produce short aliases.

File: src/lib/formatters/profile.ts [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+ * @returns Formatted row string
+ */
+export function formatProfileListRow(
+  row: ProfileFunctionRow,
+  alias?: TransactionAliasEntry
+): string {
+  const count = `${row["count()"] ?? 0}`.padStart(10);
+  const p75Ms = row["p75(function.duration)"]
+    ? formatDurationMs(row["p75(function.duration)"] / 1_000_000) // ns to ms
+    : "-";
```
[the remark this answers is not in the document]
Comment (BYK) [reply src/lib/formatters/profile.ts]: Fixed. Changed from truthiness check to explicit null/undefined check so that a `p75` value of `0` is correctly formatted instead of displaying `"-"`.

```typescript
// Before (0 treated as falsy → shows "-")
const p75Ms = row["p75(function.duration)"]
  ? formatDurationMs(...)
  : "-";

// After (only null/undefined → shows "-")
const rawP75 = row["p75(function.duration)"];
const p75Ms = rawP75 !== null && rawP75 !== undefined
  ? formatDurationMs(...)
  : "-";
```

Consistent with the `count()` field which already uses `?? 0` (nullish coalescing).
